src/storage/engine/perf_context.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::marker::PhantomData;
+
+use rocksdb::PerfContext;
+
+// TODO: Can be simplified by const generics once RFC #2000 is landed.
+
+#[derive(Clone, Debug, Copy)]
+pub struct TypeDelta;
+#[derive(Clone, Debug, Copy)]
+pub struct TypeInstant;
+
+pub trait Type: fmt::Debug {}
+impl Type for TypeDelta {}
+impl Type for TypeInstant {}
+
+/// Store statistics we need. Data comes from RocksDB's `PerfContext`.
+///
+/// Depends on different contexts, it may either store instant values (i.e. absolute values
+/// at a specific time) or delta values (i.e. changed values during some time).
+#[derive(Clone, Debug, Copy)]
+pub struct PerfStatistics<T: Type> {
```

Comment:
Hmm. This will make it more complex to access fields, we have to write "some_perf_stats.0.some_field" instead of "some_perf_stats.some_field".

Reply:
You can simply ignore the `PhantomData` here, as its name indicates. So the generic is very simple, it will produce two structs, one is XxxInstant, one is XxxDelta, whose fields are identical. The only difference is that they have different type names. In this way, we can ensure that Instant instances will not be mixed with Delta instances.

Reply:
I think it's an abuse of generic. Just define them twice, I bet it's shorten than your current implementation.

Reply:
@BusyJay So how about this implementation:

```rust
macro_rules! define_perf_statistics {
    ($($field: ident),+ $(,)*) => {
        /// Store statistics we need. Data comes from RocksDB's `PerfContext`.
        /// This statistics store instant values.
        #[derive(Debug, Clone, Copy)]
        pub struct PerfStatisticsInstant {
            $(pub $field: usize,)+
        }

        impl PerfStatisticsInstant {
            /// Create an instance which stores instant statistics values, retrieved at creation.
            pub fn new() -> Self {
                let perf_context = PerfContext::get();
                Self {
                    $($field: perf_context.$field() as usize,)+
                }
            }

            /// Calculates delta values.
            pub fn delta_between(&self, other: &Self) -> PerfStatisticsDelta {
                PerfStatisticsDelta {
                    $($field: other.$field - self.$field,)+
                }
            }

            /// Calculate delta values until now.
            pub fn delta(&self) -> PerfStatisticsDelta {
                let now = Self::new();
                self.delta_between(&now)
            }
        }

        /// Store statistics we need. Data comes from RocksDB's `PerfContext`.
        /// This this statistics store delta values between two instant statistics.
        #[derive(Default, Debug, Clone, Copy)]
        pub struct PerfStatisticsDelta {
            $(pub $field: usize,)+
        }
    }
}

define_perf_statistics!{
    internal_key_skipped_count,
    internal_delete_skipped_count,
    block_cache_hit_count,
    block_read_count,
    block_read_byte,
}
```

which completely eliminated all duplicates, and much shorter.

Reply:
LGTM, although not as readable as write them explicitly and may not be friendly to rls.